library.ml

```diff
@@ -3,64 +3,77 @@ open Printf
 open String
 open List
 
-class enemy=object (self)
-  val mutable name:string=""
-  val mutable health:int=0
-  val mutable dmg:int=0
-  method set_health n=health <- n
-  method set_name n=name <- n
-  method set_dmg n=dmg <- n
+module Enemy = struct
+  type t =
+    { name: string ;
+      health: int ;
+      dmg: int ; }
 
-  method appear:unit=
-    printf "An enemy ";print_string [red] name;printf " appeared!\n";
-    self#battle
-  method battle:unit=
-    if health>0 then begin
-      print_string [red] name;printf " attacks you for ";
-      print_string [red] (string_of_int dmg);printf " damage\n";
+  let empty = { name = "" ; health = 0 ; dmg = 0 ; }
+
+  let rec battle t : t=
+    if t.health > 0 then begin
+      print_string [red] t.name;printf " attacks you for ";
+      print_string [red] (string_of_int t.dmg);printf " damage\n";
       printf "What will you do? (";print_string [yellow] "attack";printf ")\n";
-      let response=read_line() in begin
-        printf "\n";
-        if response="attack" then (
-          health <- health+(-1);
-          printf "You did 1 damage!\n";
-          print_string [red] name;
-          if health=0 then printf " is dead!\n" else printf " has %i health left\n" health
-        ) else printf "Your attack missed\n"
-      end;
-      self#battle
-    end
+      let t =
+        let response=read_line() in begin
+          printf "\n";
+          if response="attack" then (
+            let t = {t with health = t.health - 1} in
+            printf "You did 1 damage!\n";
+            print_string [red] t.name;
+            ( if t.health=0
+              then printf " is dead!\n"
+              else printf " has %i health left\n" t.health );
+            t
+          ) else (
+            printf "Your attack missed\n" ;
+            t
+          )
+        end
+      in battle t
+    end else t
+
+  let appear t : t =
+    printf "An enemy ";print_string [red] t.name;printf " appeared!\n";
+    battle t
 end
 
-class room=object (self)
-  val mutable enemies:enemy list=[]
-  val mutable messages:string list=[]
-  val mutable rooms:room list=[]
-  val mutable name:string=""
-  method name=name
-  method set_name (n:string)=name <- n
-  method add_msg n=messages <- messages@[n]
-  method add_enemy n=enemies <- enemies@[n]
-  method add_room n=rooms <- rooms@[n]
+module Room = struct
+  type t =
+    { enemies: Enemy.t list ;
+      messages: string list ;
+      rooms: t list ;
+      name: string ; }
+  let empty = { enemies = [] ; messages = [] ; rooms = [] ;name = ""; }
+  let add_msg n t = {t with messages = t.messages@[n]}
+  let add_enemy n t= {t with enemies = t.enemies@[n]}
+  let add_room n ~parent:({rooms;_} as t) = {t with rooms = rooms@[n]}
+
+  let rec choose_room self :t=
+    let input=read_line() in printf "\n";
+    match List.find_opt (fun a -> equal a.name input) self.rooms with
```

Comment:
Instead of constructing a new list with `List.find_all` we use `List.find_opt` to obtain a `option`.
The option type is defined like this:
```ocaml
type 'a option =
  | None
  | Some of 'a
```

When an element is not found, we get a `None` value.
When an element **is** found, we get `Some element` that we can pattern match on.

`List.find_opt` instead of `List.find_all` avoids iterating through *all* the elements in the list since it will return after the first matching element is found.